compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2714,76 +2715,193 @@ public void visit(BLangInvocation iExpr) {
+            for (BLangNamedArgsExpression namedArg : namedArgs) {
+                BField field = targetFields.get(namedArg.name.value);
+                Location pos = namedArg.pos;
+                if (field == null) {
+                    if (targetErrorDetailRec.sealed) {
+                        dlog.error(pos, DiagnosticErrorCode.UNKNOWN_DETAIL_ARG_TO_CLOSED_ERROR_DETAIL_REC,
+                                namedArg.name, targetErrorDetailRec);
+                    } else if (targetFields.isEmpty()
+                            && !types.isAssignable(namedArg.expr.type, targetErrorDetailRec.restFieldType)) {
+                        dlog.error(pos, DiagnosticErrorCode.INVALID_ERROR_DETAIL_REST_ARG_TYPE,
+                                namedArg.name, targetErrorDetailRec);
+                    }
+                } else {
+                    missingRequiredFields.remove(namedArg.name.value);
+                    if (!types.isAssignable(namedArg.expr.type, field.type)) {
+                        dlog.error(pos, DiagnosticErrorCode.INVALID_ERROR_DETAIL_ARG_TYPE,
+                                namedArg.name, field.type, namedArg.expr.type);
+                    }
+                }
             }
-        } else if (detailType.tag == TypeTags.RECORD) {
-            BRecordType targetErrorDetailRec = (BRecordType) errorConstructorType.detailType;
-            BRecordType recordType = createErrorDetailRecordType(errorConstructorExpr, targetErrorDetailRec);
-            if (resultType == symTable.semanticError) {
-                return;
+
+            for (String requiredField : missingRequiredFields) {
+                dlog.error(errorConstructorExpr.pos, DiagnosticErrorCode.MISSING_ERROR_DETAIL_ARG, requiredField);
             }
+        }
 
-            if (!types.isAssignable(recordType, targetErrorDetailRec)) {
-                reportErrorDetailMissmatchError(errorConstructorExpr, targetErrorDetailRec, recordType);
-                resultType = symTable.semanticError;
-                return;
+        errorConstructorExpr.type = errorType;
+        resultType = types.checkType(errorConstructorExpr.pos, errorType, expType,
+                DiagnosticErrorCode.INCOMPATIBLE_TYPES);
+    }
+
+    private void validateErrorConstructorPositionalArgs(BLangErrorConstructorExpr errorConstructorExpr) {
+        // Parser handle the missing error message case, and too many positional argument cases.
+        if (errorConstructorExpr.positionalArgs.isEmpty()) {
+            return;
+        }
+
+        checkExpr(errorConstructorExpr.positionalArgs.get(0), this.env, symTable.stringType);
+
+        int positionalArgCount = errorConstructorExpr.positionalArgs.size();
+        if (positionalArgCount > 1) {
+            checkExpr(errorConstructorExpr.positionalArgs.get(1), this.env, symTable.errorOrNilType);
+        }
+
+        // todo: Need to add type-checking when fixing #29247 for positional args beyond second arg.
```

Comment:
Shall we create a separate issue for this?

Reply:
Shall we make this a part of #29247